### PR TITLE
fix(nvmx): controller i/o statistics

### PR DESCRIPTION
**Pull request by mtzaurus**
This fix introduces I/O statistics for read, write and unmap
operations for NVMe block devices.

Implements CAS-722

### CODE REVIEW

**Review by gila — APPROVED** [review]
👍

File: mayastor/src/bdev/dev/nvmx/controller.rs [hunk 1]
```diff
@@ -328,6 +330,70 @@ impl<'a> NvmeController<'a> {
...
     }
 
+    /// Get I/O statistics for all I/O channels of the controller.
+    pub fn get_io_stats<T: 'static + Sized, F>(
+        &self,
+        cb: F,
+        cb_arg: T,
+    ) -> Result<(), CoreError>
+    where
+        F: Fn(Result<BlockDeviceIoStats, CoreError>, T) + 'static,
```
Comment (jkryl) [inline mayastor/src/bdev/dev/nvmx/controller.rs:340]: just an idea. By pushing the oneshot channel from io_stats() to this routine, you could keep larger part of the rust code clean (async method without without callbacks). But maybe it cannot be done for some reason.

File: mayastor/src/bdev/dev/nvmx/channel.rs [hunk 2]
```diff
@@ -291,6 +295,61 @@ impl NvmeIoChannelInner<'_> {
...
+
+    #[inline]
+    /// Account amount of blocks and I/O operations.
+    pub fn account_block_io(
+        &mut self,
+        op: IoType,
+        num_ops: u64,
+        num_blocks: u64,
+    ) {
+        let num_bytes = num_blocks * self.block_size;
```
Comment (jkryl) [inline mayastor/src/bdev/dev/nvmx/channel.rs:327]: Not sure how much raw the stats should be. But I personally would prefer as raw numbers as possible and I can do the remaining math as a consumer of the stats. Also the code in IO path that is called often, could be simpler then. So if we recorded just number of reads and writes we could save a couple of instructions and a couple of bytes of memory. Assuming the consumer knows the block size of the device, she can calculate the rest easily. Just an idea and it's really a micro optimisation.

**Review by jkryl — APPROVED** [review]
👍

File: mayastor/src/bdev/dev/nvmx/channel.rs (hunk 2 above)
Comment (mtzaurus) [reply mayastor/src/bdev/dev/nvmx/channel.rs:327]: Actually, libSPDK does the same as we do, they calculate the real stats in place:

bdev_io->internal.ch->stat.bytes_read += bdev_io->u.bdev.num_blocks * bdev_io->bdev->blocklen;

Looks like they don't care much about the speed. :-)

But not a problem, I can move multiplication step to 'get_stats()' and return updated stats when explicitly requested. I would prefer to keep translation as close to the stats itself as possible and not to put any burden to the caller.

Comment (jonathan-teh) [reply mayastor/src/bdev/dev/nvmx/channel.rs:327]: It appears that all callers pass 1 to `num_ops` so perhaps that can be dropped.

File: mayastor/src/bdev/dev/nvmx/controller.rs (hunk 1 above)
Comment (mtzaurus) [reply mayastor/src/bdev/dev/nvmx/controller.rs:340]: Unfortunately, we can't use this method as async. We invoke controller's methods under protection of the mutex, and await()'ing under mutex is not always a good idea. In fact, our linter complains about such a thing. That's why we initiate an async operation and then wait with the controller freed.

File: mayastor/src/bdev/dev/nvmx/device.rs [hunk 3]
```diff
@@ -173,8 +176,28 @@ impl BlockDevice for NvmeBlockDevice {
...
+            CoreError::BdevNotFound {
+                name: self.name.to_string(),
+            },
+        )?;
+
+        let (s, r) =
+            oneshot::channel::<Result<BlockDeviceIoStats, CoreError>>();
+        // Schedule async I/O stats collection and wait for the result.
+        {
+            let controller = carc.lock().expect("lock poisened");
```
Comment (jonathan-teh) [inline mayastor/src/bdev/dev/nvmx/device.rs:190]: Typo: poisoned.

**Review by jonathan-teh — APPROVED** [review]

File: mayastor/src/bdev/dev/nvmx/device.rs (hunk 3 above)
Comment (mtzaurus) [reply mayastor/src/bdev/dev/nvmx/device.rs:190]: Typo fixed, thanks.

**Pull request merged**